### PR TITLE
fix(compiler): improve error message when a module imports itself

### LINKED ISSUE #14644: Better error message when module imports itself
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

When a module imports itself a generic 
```Uncaught RangeError: Maximum call stack size exceeded```
is thrown which could have a whole bunch of reasons.

**Expected behavior**

It would be really nice if Angular could detect self-imports (maybe even circular imports if that's a thing) and show a more helpful error message to spare developers the time of debugging.

**Minimal reproduction of the problem with instructions**

http://plnkr.co/edit/hD9KHmgNKgiNYwXCUYp4

Run and check the console.
Removing AppModule from AppModule's imports resolves the problem.

**What is the motivation / use case for changing the behavior?**

Cause could be a developers mistake, wrong entry of autocomplete (many IDEs offer it as an option since the class is in scope) or even refactoring gone wrong (which would possibly cause hours of debugging, searching for self-generated loops in functions etc.), ...

Best clue when searching on the internet is this content, which is not even really related to the issue topic
https://github.com/angular/angular/issues/7930#issuecomment-259699938
* **Angular version:** 2.4.8

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Closes #14644

### CODE REVIEW

**Review by chuckjaz — APPROVED** [review]
I like this change. Do we also want to check for indirect cycles as well (e.g. `A` imports `B` which imports 'A').

**Comment by DzmitryShylovich** [comment]
> Do we also want to check for indirect cycles as well

I think it's a good idea. I'll create a new issue for it

**Comment by chuckjaz** [comment]
Note to CARETAKER, this fix can be merged independently of #14649 being fixed.

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 1]
```diff
@@ -638,6 +639,18 @@ export class CompileMetadataResolver {
     return compileMeta;
   }
 
+  private _checkSelfImport(moduleType: Type<any>, importedModuleType: Type<any>): boolean {
+    if (moduleType === importedModuleType) {
+      const moduleName: string = stringifyType(moduleType);
+      this._reportError(
+          syntaxError(
+              `Unexpected module '${moduleName}' imported by the module '${moduleName}'. NgModule can't import itself`),
```
Comment (vicb) [inline modules/@angular/compiler/src/metadata_resolver.ts]: What about `'${moduleName}' can't import itself` - I don't think the first sentence improves clarity here ?

Comment (DzmitryShylovich) [inline modules/@angular/compiler/src/metadata_resolver.ts]: yeah, it's much simpler.
fixed

**Pull request merged**